src/headless_browser.ts

```diff
@@ -58,82 +67,215 @@ export type ExceptionDetails = { // exists when an error
   text: string; // eg Uncaught
 };
 
-export type DOMOutput = {
+type DOMOutput = {
   result: SuccessResult | ErrorResult | UndefinedResult;
   exceptionDetails?: ExceptionDetails; // exists when an error, but an undefined response value wont trigger it, for example if the command is `window.loction`, there is no `exceptionnDetails` property, but if the command is `window.` (syntax error), this prop will exist
 };
 
+const webSocketIsDonePromise = deferred();
+
 export class HeadlessBrowser {
   /**
    * The sub process that runs headless chrome
    */
-  private readonly browser_process: Deno.Process;
+  private browser_process: Deno.Process | null = null;
+
+  /**
+   * Our web socket connection to the remote debugging port
+   */
+  private socket: WebSocket | null = null;
 
   /**
-   * A way to keep track of the last command sent.
-   * Used to display the command in the error message if
-   * command errored
+   * A counter that acts as the message id we use to send as part of the event data through the websocket
    */
-  private last_command_sent = "";
+  private next_message_id = 1;
 
   /**
-   * A list of every command sent
+   * To keep hold of promises waiting for a notification from the websocket
    */
-  private commands_sent: string[] = [];
+  private notification_resolvables: { [key: string]: Deferred<void> } = {};
 
   /**
-   * @param urlToVisit - The url to visit or open up
+   * To keep hold of our promises waiting for messages from the websocket
    */
-  constructor(urlToVisit: string) {
+  private resolvables: { [key: number]: Deferred<unknown> } = {};
+
+  constructor() {
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - PUBLIC ////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Build the headless browser
+   */
+  public async build() {
+    const paths = {
+      // deno-lint-ignore camelcase
+      windows_chrome_exe:
+        "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
+      // deno-lint-ignore camelcase
+      windows_chrome_exe_x86:
+        "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
+      darwin: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
+      linux: "/usr/bin/google-chrome",
+    };
     let chromePath = "";
     switch (Deno.build.os) {
       case "darwin":
-        chromePath =
-          "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";
+        chromePath = paths.darwin;
         break;
       case "windows":
-        chromePath = "start chrome";
-        break;
+        if (existsSync(paths.windows_chrome_exe)) {
```

Comment:
Can we either:
1. Rewrite the entire `existsSync` function to be asynchronous?
2. In case that we use `existsSync` in a place that we cannot `await` it, maybe have an `exists` that is asynchronous and use it here?

Reply:
nitpicking is good, and good point, makes sense to use async: https://github.com/drashland/sinco/pull/19/commits/72e6566107044e5f8f0741cab9d2aa87bd076afd